internal/controllers/admission_test.go

```diff
@@ -19,159 +19,178 @@ func operator(spec operatorsv1alpha1.OperatorSpec) *operatorsv1alpha1.Operator {
-			})
-			err := cl.Create(ctx, op)
-			Expect(err).NotTo(HaveOccurred(), "expected success for semver range '%q': %w", validSemver, err)
-			err = cl.Delete(ctx, op)
-			Expect(err).NotTo(HaveOccurred(), "unexpected error deleting valid semver '%q': %w", validSemver, err)
-		}
-	})
-	It("should fail if an invalid channel name is given", func() {
-		invalidChannels := []string{
-			"spaces spaces",
-			"Capitalized",
-			"camelCase",
-			"many/invalid$characters+in_name",
-			"-start-with-hyphen",
-			"end-with-hyphen-",
-			".start-with-period",
-			"end-with-period.",
-		}
-		for _, invalidChannel := range invalidChannels {
-			err := cl.Create(ctx, operator(operatorsv1alpha1.OperatorSpec{
-				PackageName: "package",
-				Channel:     invalidChannel,
-			}))
-			Expect(err).To(HaveOccurred(), "expected error for invalid channel '%q'", invalidChannel)
-			Expect(err.Error()).To(ContainSubstring("spec.channel in body should match '^[a-z0-9]+([\\.-][a-z0-9]+)*$'"))
-		}
-	})
-	It("should pass if a valid channel name is given", func() {
-		validChannels := []string{
-			"hyphenated-name",
-			"dotted.name",
-			"channel-has-version-1.0.1",
-		}
-		for _, validChannel := range validChannels {
-			op := operator(operatorsv1alpha1.OperatorSpec{
-				PackageName: "package",
-				Channel:     validChannel,
-			})
-			err := cl.Create(ctx, op)
-			Expect(err).NotTo(HaveOccurred(), "unexpected error creating valid channel '%q': %w", validChannel, err)
-			err = cl.Delete(ctx, op)
-			Expect(err).NotTo(HaveOccurred(), "unexpected error deleting valid channel '%q': %w", validChannel, err)
-		}
-	})
-	It("should fail if an invalid channel name length", func() {
+		require.Errorf(t, err, "expected error for invalid semver %q", invalidSemver)
+		// Don't need to include the whole regex, this should be enough to match the MasterMinds regex
+		require.ErrorContains(t, err, "spec.version in body should match '^(\\s*(=||!=|>|<|>=|=>|<=|=<|~|~>|\\^)")
+	}
+}
+
+func TestOperatorValidSemver(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	validSemvers := []string{
+		">=1.2.3",
+		"=>1.2.3",
+		">= 1.2.3",
+		">=v1.2.3",
+		">= v1.2.3",
+		"<=1.2.3",
+		"=<1.2.3",
+		"=1.2.3",
+		"!=1.2.3",
+		"<1.2.3",
+		">1.2.3",
+		"~1.2.2",
+		"~>1.2.3",
+		"^1.2.3",
+		"1.2.3",
+		"v1.2.3",
+		"1.x",
+		"1.X",
+		"1.*",
+		"1.2.x",
+		"1.2.X",
+		"1.2.*",
+		">=1.2.3 <2.3.4",
+		">=1.2.3,<2.3.4",
+		">=1.2.3, <2.3.4",
+		"<1.2.3||>2.3.4",
+		"<1.2.3|| >2.3.4",
+		"<1.2.3 ||>2.3.4",
+		"<1.2.3 || >2.3.4",
+		">1.0.0,<1.2.3 || >2.1.0",
+		"<1.2.3-abc >2.3.4-def",
+		"<1.2.3-abc+def >2.3.4-ghi+jkl",
+	}
+	for _, validSemver := range validSemvers {
+		op := operator(operatorsv1alpha1.OperatorSpec{
+			PackageName: "package",
+			Version:     validSemver,
+		})
+		err := cl.Create(ctx, op)
+		require.NoErrorf(t, err, "unexpected error for semver range '%q': %w", validSemver, err)
+		err = cl.Delete(ctx, op)
```

Comment:
If this is an integration test, the testenv process is killed at the end of tests for the package. I don't think there is any value to doing "cleanup" - it takes time that is not rewarded.

Reply:
Again, direct translation, optimizations later.

Reply:
Strike that last comment. In this case, while the `testEnv` is terminated at the end of the package, the operators created during the success cases remain until deleted, and subsequent tests fail because those operators exist; this is because the same name is used. The choice is between randomizing the name, or deleting the operator resource.

Reply:
Randomize the name and run every test in parallel, every time.

Reply:
The last commit adds a random name, I'd have to look at making it parallel.

Reply:
It saves _maybe_ 1/2 of a second on a 7~8 second test. Any savings is lost in the variability in the test's duration.